src/NuGet.Core/NuGet.Protocol.Core.v3/PackageUpdateResource.cs

```diff
@@ -0,0 +1,601 @@
+﻿using System;
+using System.IO;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using NuGet.Configuration;
+using NuGet.Logging;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.Versioning;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.RegularExpressions;
+using NuGet.Protocol.Core.v3;
+using System.Globalization;
+
+
+namespace NuGet.Protocol.Core.Types
+{
+    /// <summary>
+    /// Contains logics to push or delete packages in Http server or file system
+    /// </summary>
+    public class PackageUpdateResource : INuGetResource
+    {
+        private const string ServiceEndpoint = "/api/v2/package";
+        private const string ApiKeyHeader = "X-NuGet-ApiKey";
+
+        private HttpSource _httpSource;
+        private string _source;
+
+        public PackageUpdateResource(string source,
+            HttpSource httpSource)
+        {
+            _source = source;
+            _httpSource = httpSource;
+        }
+
+        public async Task Push(string packagePath,
+            string source,
+            int timeoutInSecond,
+            Func<string, string> getApiKey,
+            ILogger log)
+        {
+            var timeout = TimeSpan.FromSeconds(Math.Abs(timeoutInSecond));
+            if (timeout.TotalSeconds == 0)
```

Comment:
Normally zero or -1 means no timeout, there should be a way to provide that here instead of passing in MaxInt